tests/components/http/test_forwarded.py

```diff
@@ -0,0 +1,487 @@
+
+
+async def test_x_forwarded_proto_without_trusted_proxy(aiohttp_client):
+    """Test that proto header is ignored when untrusted."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "http"
+        assert not request.secure
+        assert request.remote == "127.0.0.1"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+
+    async_setup_forwarded(app, [])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/", headers={X_FORWARDED_FOR: "255.255.255.255", X_FORWARDED_PROTO: "https"}
+    )
+
+    assert resp.status == 200
+
+
+@pytest.mark.parametrize(
+    "x_forwarded_for,remote,x_forwarded_proto,secure",
+    [
+        ("10.10.10.10, 127.0.0.1, 127.0.0.2", "10.10.10.10", "https, http, http", True),
+        ("10.10.10.10, 127.0.0.1, 127.0.0.2", "10.10.10.10", "https,http,http", True),
+        ("10.10.10.10, 127.0.0.1, 127.0.0.2", "10.10.10.10", "http", False),
+        (
+            "10.10.10.10, 127.0.0.1, 127.0.0.2",
+            "10.10.10.10",
+            "http, https, https",
+            False,
+        ),
+        ("10.10.10.10, 127.0.0.1, 127.0.0.2", "10.10.10.10", "https", True),
+        (
+            "255.255.255.255, 10.10.10.10, 127.0.0.1",
+            "10.10.10.10",
+            "http, https, http",
+            True,
+        ),
+        (
+            "255.255.255.255, 10.10.10.10, 127.0.0.1",
+            "10.10.10.10",
+            "https, http, https",
+            False,
+        ),
+        ("255.255.255.255, 10.10.10.10, 127.0.0.1", "10.10.10.10", "https", True),
+    ],
+)
+async def test_x_forwarded_proto_with_trusted_proxy(
+    x_forwarded_for, remote, x_forwarded_proto, secure, aiohttp_client
+):
+    """Test that we get the proto header if proxy is trusted."""
+
+    async def handler(request):
+        assert request.remote == remote
+        assert request.scheme == ("https" if secure else "http")
+        assert request.secure == secure
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.0/24")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/",
+        headers={
+            X_FORWARDED_FOR: x_forwarded_for,
+            X_FORWARDED_PROTO: x_forwarded_proto,
+        },
+    )
+
+    assert resp.status == 200
+
+
+async def test_x_forwarded_proto_with_trusted_proxy_multiple_for(aiohttp_client):
+    """Test that we get the proto with 1 element in the proto, multiple in the for."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "https"
+        assert request.secure
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)